projects/npm-tools/packages/npm-scripts/test/typescript/getTypeScriptBuildOrder.js

```diff
@@ -0,0 +1,234 @@
+/**
+ * SPDX-FileCopyrightText: © 2019 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+const getTypeScriptBuildOrder = require('../../src/typescript/getTypeScriptBuildOrder');
+
+describe('getTypeScriptBuildOrder()', () => {
+	it('handles an empty graph', () => {
+		expect(getTypeScriptBuildOrder({})).toEqual([]);
+	});
+
+	it('handles a graph with one item', () => {
+		expect(
+			getTypeScriptBuildOrder({
+				'@liferay/remote-app-client-js': {
+					dependencies: {},
+					name: '@liferay/remote-app-client-js',
+				},
+			})
+		).toEqual([
+			{
+				dependencies: {},
+				name: '@liferay/remote-app-client-js',
+			},
+		]);
+	});
+
+	it('handles a graph with two items and no dependency', () => {
+		expect(
+			getTypeScriptBuildOrder({
+				'@liferay/remote-app-client-js': {
+					dependencies: {},
+					name: '@liferay/remote-app-client-js',
+				},
+				'@liferay/some-other-module': {
+					dependencies: {},
+					name: '@liferay/some-other-module',
+				},
+			})
+		).toEqual([
```

Comment:
Is this stable in that it always returns in the same order or depends on the input? Is there a strong reason for one thing or the other?

Reply:
Some examples are so trivial as to be stable no matter what. This one, concretely, depends on input order. As you move to more complicated graphs (even this little one), you get stable output unless you change the order the dependencies are declared, or the order in which you list the nodes in graph, _but_ the output is guaranteed to indicate a valid topological ordering.

This all stems from the fact that there can be more than one valid topological orderings of a graph. The simplest example that is actually interesting that illustrates this is the "diamond-shaped" graph seen elsewhere in this test file:

```
         C
        / \        ^
       A   D       | dependencies point upwards
        \ /        |
         B
```

`C` _must_ come first, then it doesn't matter what order you build `A` + `D` in, and `B` can only happen after all the others are done. So the two valid orderings are `[C, A, D, B]` and `[C, D, A, B]`. Which one you produce depends on implementation details + the ordering you pass in, as I mentioned above.

> Is there a strong reason for one thing or the other?

If we really cared about stability, I think we could guarantee it by enforcing input order (ie. that the nodes in the graph were passed in in lexicographical order, and that they declared their dependencies in lexicographical order too).

Reply:
Sounds good! 👍